src/main/java/net/starschema/clouddb/jdbc/BQStatementRoot.java

```diff
@@ -240,6 +241,73 @@ public boolean execute(String arg0, String[] arg1) throws SQLException {
         throw new BQSQLException("Not implemented." + "executeBatch()");
     }
 
+    /**
+     * Execute DML (DELETE, INSERT, UPDATE). If you want to perform a SELECT or other DDL
+     * look at {@link #execute(String)}
+     */
+    private int executeDML(String sql) throws SQLException {
+        if (this.isClosed()) {
+            throw new BQSQLException("This Statement is Closed");
+        }
+        if (!connection.shouldUseQueryApi()) {
+            throw new BQSQLFeatureNotSupportedException("Only Query API is supported");
+        }
+        if (connection.getUseLegacySql()) {
+            throw new BQSQLFeatureNotSupportedException("Legacy SQL does not support DML");
+        }
+        this.starttime = System.currentTimeMillis();
+        // ANTLR Parsing
+        BQQueryParser parser = new BQQueryParser(sql, this.connection);
+        sql = parser.parse();
+
+        final Job referencedJob;
+
+        try {
+            QueryResponse qr = BQSupportFuncts.runSyncQuery(
+                    this.connection.getBigquery(),
+                    this.ProjectId,
+                    sql,
+                    connection.getDataSet(),
+                    this.connection.getUseLegacySql(),
+                    this.connection.getMaxBillingBytes(),
+                    (long) querytimeout * 1000,
+                    (long) getMaxRows()
+            );
+
+            if (qr.getJobComplete()) {
+                // I hope they don't insert more than 2^32-1 :)
+                return Math.toIntExact(qr.getNumDmlAffectedRows());
+            }
+
+            referencedJob = this.connection.getBigquery().jobs().get(this.ProjectId, qr.getJobReference().getJobId()).execute();
+        } catch (IOException e) {
+            throw new BQSQLException("Something went wrong with the query: " + sql, e);
+        }
+
+        try {
+            do {
+                if (BQSupportFuncts.getQueryState(referencedJob,
+                        this.connection.getBigquery(), this.ProjectId).equals(
+                        "DONE")) {
+                    return Math.toIntExact(
+                                BQSupportFuncts.getQueryResults(
+                                        this.connection.getBigquery(), this.ProjectId,
+                                        referencedJob).getNumDmlAffectedRows()
+                            );
+                }
+                Thread.sleep(500);
+                this.logger.debug("slept for 500" + "ms, querytimeout is: "
+                        + this.querytimeout + "s");
+            } while (System.currentTimeMillis() - this.starttime <= (long) this.querytimeout * 1000);
+        } catch (IOException | InterruptedException e) {
+            throw new BQSQLException("Something went wrong with the query: " + sql, e);
+        }
```

Comment:
TODO: Probably can wrap this functionality as it's used a ton in the codebase into a helper method.

Reply:
This is a TODO for later. This pattern is used a little bit and should be cleaned up:
```bash
❯ rg "while \(System.currentTimeMillis\(\)"
src/main/java/net/starschema/clouddb/jdbc/BQPreparedStatement.java
283:            while (System.currentTimeMillis() - this.starttime <= (long) this.querytimeout * 1000);

src/main/java/net/starschema/clouddb/jdbc/BQStatementRoot.java
298:            } while (System.currentTimeMillis() - this.starttime <= (long) this.querytimeout * 1000);
385:            while (System.currentTimeMillis() - this.starttime <= (long) this.querytimeout * 1000);

src/main/java/net/starschema/clouddb/jdbc/BQStatement.java
223:            while (System.currentTimeMillis() - this.starttime <= (long) this.querytimeout * 1000);
```

This JDBC driver in fact can use _a lot of love_ especially given that the state of the BigQuery API has changed since 2015. It might make sense to make a branch that removes legacy features as well and have our standard SQL use that instead.

(A more interesting idea would be to either migrate to the Samba JDBC driver or I think we could write one ourselves better suited for our needs)